File: oo/src/test/java/oo/ep/M1Test.java
package oo.ep;
import org.junit.Assert;
import org.junit.Test;

public class M1Test {

    public interface TestTemplate extends M0Test.TestTemplate {
        default void test() {
            M0Test.TestTemplate.super.test();

            Exp expr1 = new Sub(new Lit(1.0), new Lit(2.0));
            Assert.assertEquals(-1.0, expr1.eval(), 0.0);

            Exp expr2 = new Lit(2.0);
            Assert.assertEquals(2.0, expr2.eval(), 0.0);
            Assert.assertEquals(new Sub(new Lit(1.0), new Lit(2.0)).eval(), -1.0, 0.0);
        }
    }
    private static class ActualTest implements TestTemplate {}

    @Test
    public void testTest() { new ActualTest().test(); }
}

File: coco_scala/src/test/scala/exp/m1/M1Test.scala
package exp.m1

import org.scalatest.FunSuite

trait TestTemplate[FT] extends Factory[FT] with exp.m0.TestTemplate[FT] {
  val suite:FunSuite
  import suite._

  override def test() : Unit = {
    super.test()

    val expr1 = this.sub(this.lit(1.0), this.lit(2.0))
    assert(-1.0 === expr1.eval)

    val expr2 = this.lit(2.0)
    assert(2.0 === expr2.eval)
    assert(this.sub(this.lit(1.0), this.lit(2.0)).eval === -1.0)
  }
}

class M1Test extends FunSuite { self =>

  object ActualTest extends TestTemplate[exp.m0.finalized.Exp] with finalized.Factory {
    val suite: FunSuite = self
  }

  test("M1") {
    ActualTest.test()
  }
}
